pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,138 @@
+package kubeletversionskewcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+const (
+	KubeletVersionSkewLimitUpgradeableConditionType = "KubeletVersionSkewLimitUpgradeable"
+
+	KubeletVersionUnknownReason           = "KubeletVersionUnknown"
+	KubeletVersionSyncedReason            = "KubeletVersionSynced"
+	KubeletVersionSkewLessThanLimitReason = "KubeletVersionSkewLessThanLimit"
+	KubeletVersionSkewAtLimitReason       = "KubeletVersionSkewAtLimit"
+	KubeletVersionUnsupportedReason       = "KubeletVersionUnsupported"
+)
+
+// KubeletVersionSkewController sets Upgradeable=False if the kubelet
+// version on a node falls too far behind the api server version.
+type KubeletVersionSkewController interface {
+	factory.Controller
+}
+
+func NewKubeletVersionSkewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:   operatorClient,
+		nodeLister:       kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion: semver.MustParse(status.VersionForOperandFromEnv()),
```

Comment:
Comes from the operator's (in this case kas-o) `OPERAND_IMAGE_VERSION`  environment variable which is in turn updated by the CVO(?).

Reply:
actually via `oc`.  Because the operator's deployment manifest has some [`0.0.1-snapshot` placeholders][1], and [those are replaced with the version string by `oc` at release-image-build time][2] ([code][3]).

[1]: https://github.com/openshift/cluster-openshift-apiserver-operator/blob/4da70a9e816206146c43bd4de0d6f7655a52d9a0/manifests/0000_30_openshift-apiserver-operator_07_deployment.yaml#L54-L57
[2]: https://github.com/openshift/enhancements/blob/f97876821d3bb506d28fee565271d3bebbbc682c/dev-guide/cluster-version-operator/dev/clusteroperator.md#what-should-be-the-contents-of-clusteroperator-custom-resource-in-manifests
[3]: https://github.com/openshift/oc/blob/f94afb52dc8a3185b3b9eacaf92ec34d80f8708d/pkg/cli/admin/release/image_mapper.go#L324-L334

Reply:
@wking out of curiosity: How and where do those `0.0.1-snapshot` mappings translate to environment variables? I think the `OPERATOR_IMAGE_VERSION` env var is derived from the constant name `operator` in https://github.com/openshift/cluster-openshift-apiserver-operator/blob/4da70a9e816206146c43bd4de0d6f7655a52d9a0/manifests/0000_30_openshift-apiserver-operator_08_clusteroperator.yaml#L12-L13

But how is `OPERAND_IMAGE_VERSION` being derived if theoretically multiple operand versions can be specified https://github.com/openshift/cluster-openshift-apiserver-operator/blob/4da70a9e816206146c43bd4de0d6f7655a52d9a0/manifests/0000_30_openshift-apiserver-operator_08_clusteroperator.yaml#L14-L15 ?

I didn't find the code base for setting those env vars, do you have a pointer?

Reply:
@s-urbaniak It's just a simple string replacement. In your example, it is used twice since the versions there should always be in sync and match the overall OCP release version. Another common replacement is for `1.0.0-snapshot-kubernetes` which corresponds to the upstream `kubernetes` "component" version. I think the release version and the various component versions (of which I only know of `kubernetes`) are extracted from the release's ImageStream. Hopefully @wking can clarify.

Reply:
It's not my code, so I don't have a complete handle on this, but the godocs for [the code I linked earlier][1] seem reasonably clear.  They say `0.0.1-snapshot-[component]` though, so it would be `0.0.1-snapshot-kubernetes`, etc., not `1.0.0-snapshot-kubernetes`.  [Looks][2] [like][3] the component versions are pulled out of a `io.openshift.build.versions` annotation in the ImageStream feeding the release image (and maybe from the image config for stuff that's not sucked in off an ImageStream?).  Spot-checking 4.8.5:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-upgrade/1425332773236772864/artifacts/release/artifacts/release-images-latest | jq -r '.spec.tags[] | .annotations["io.openshift.build.versions"] as $v | select($v != null) | .name + " " + $v'
hyperkube kubernetes=1.21.1
machine-os-content machine-os=48.84.202108062347-0
```

[1]: https://github.com/openshift/oc/blob/f94afb52dc8a3185b3b9eacaf92ec34d80f8708d/pkg/cli/admin/release/image_mapper.go#L324-L334
[2]: https://github.com/openshift/oc/blob/f94afb52dc8a3185b3b9eacaf92ec34d80f8708d/pkg/cli/admin/release/image_mapper.go#L174-L188
[3]: https://github.com/openshift/oc/blob/f94afb52dc8a3185b3b9eacaf92ec34d80f8708d/pkg/cli/admin/release/annotations.go#L19-L24